File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/consent/scopes.rb
# frozen_string_literal: true

module IAIso
  module Consent
    # Scope-matching logic.
    module Scopes
      module_function

      # Returns true iff the requested scope is granted by any in `granted`.
      # Match rules per spec/consent/README.md:
      #   - Exact match: granted "tools.search" satisfies requested "tools.search".
      #   - Prefix-at-segment-boundary: granted "tools" satisfies "tools.search".
      #   - Substring without boundary does NOT match: "tools" vs "toolsbar".
      #   - More-specific does NOT satisfy less-specific: "tools.search.bulk"
      #     does not satisfy "tools.search".
      #
      # Raises ArgumentError if requested is empty.
      def granted(granted, requested)
        raise ArgumentError, "requested scope must be non-empty" if requested.nil? || requested.empty?
        Array(granted).any? do |g|
          g == requested || requested.start_with?("#{g}.")
        end
      end
    end
  end
end

File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoConsent/Scopes.swift
import Foundation

/// Scope-matching logic.
public enum Scopes {

    /// Returns true iff `requested` is granted by any scope in `granted`.
    /// Match rules per `spec/consent/README.md`:
    ///
    /// - Exact match: granted "tools.search" satisfies requested "tools.search".
    /// - Prefix-at-segment-boundary: granted "tools" satisfies requested "tools.search"
    ///   (boundary is the dot).
    /// - Substring without boundary does NOT match: "tools" does not satisfy "toolsbar".
    /// - More-specific does NOT satisfy less-specific: "tools.search.bulk" does
    ///   not satisfy "tools.search".
    ///
    /// - Throws: `ScopesError.emptyRequested` if `requested` is empty.
    public static func granted(_ granted: [String], _ requested: String) throws -> Bool {
        if requested.isEmpty {
            throw ScopesError.emptyRequested
        }
        for g in granted {
            if g == requested { return true }
            if requested.hasPrefix(g + ".") { return true }
        }
        return false
    }
}

public enum ScopesError: Error, CustomStringConvertible, Sendable {
    case emptyRequested
    public var description: String {
        switch self {
        case .emptyRequested: return "requested scope must be non-empty"
        }
    }
}
